kf_task_fhir_etl/etl/ingest.py

```diff
@@ -0,0 +1,525 @@
+                logging.info(f"    📁 {endpoint} {df.shape}")
+
+            logging.info(f"  ✅ Extracted {kf_study_id}")
+
+        return mapped_df_dict
+
+    def transform(self, mapped_df_dict):
+        """Transforms records.
+
+        :param mapped_df_dict: An output from the above exract stage
+        :type mapped_df_dict: dict
+        :return: A dictionary of outer-merged data frames
+        :rtype: dict
+        """
+        merged_df_dict = defaultdict()
+
+        for kf_study_id, study_mapped_df_dict in mapped_df_dict.items():
+            logging.info(f"  ⏳ Transforming {kf_study_id}")
+            merged_df_dict.setdefault(kf_study_id, {})
+            study_merged_df, study_all_targets = None, set()
+
+            # studies
+            studies = study_mapped_df_dict["studies"]
+            studies = studies.rename(
+                columns={
+                    "investigator_id": CONCEPT.INVESTIGATOR.TARGET_SERVICE_ID,
+                    "attribution": CONCEPT.STUDY.ATTRIBUTION,
+                    "data_access_authority": CONCEPT.STUDY.AUTHORITY,
+                    "domain": "STUDY|DOMAIN",
+                    "external_id": CONCEPT.STUDY.ID,
+                    "kf_id": CONCEPT.STUDY.TARGET_SERVICE_ID,
+                    "name": CONCEPT.STUDY.NAME,
+                    "program": "STUDY|PROGRAM",
+                    "release_status": CONCEPT.STUDY.RELEASE_STATUS,
+                    "short_code": "STUDY|SHORT_CODE",
+                    "short_name": CONCEPT.STUDY.SHORT_NAME,
+                    "version": CONCEPT.STUDY.VERSION,
+                    "visible": CONCEPT.STUDY.VISIBLE,
+                }
+            )
+            study_all_targets.add(ResearchStudy)
+
+            # investigators
+            investigators = study_mapped_df_dict.get("investigators")
+            if investigators is not None:
+                investigators = investigators.rename(
+                    columns={
+                        "external_id": CONCEPT.INVESTIGATOR.ID,
+                        "institution": CONCEPT.INVESTIGATOR.INSTITUTION,
+                        "kf_id": CONCEPT.INVESTIGATOR.TARGET_SERVICE_ID,
+                        "name": CONCEPT.INVESTIGATOR.NAME,
+                        "visible": CONCEPT.INVESTIGATOR.VISIBLE,
+                    }
+                )
+                study_merged_df = outer_merge(
+                    studies,
+                    investigators,
+                    with_merge_detail_dfs=False,
+                    on=CONCEPT.INVESTIGATOR.TARGET_SERVICE_ID,
+                )
+                study_all_targets.update(
+                    [
+                        Practitioner,
+                        Organization,
+                        PractitionerRole,
+                    ]
+                )
+
+            # participants
+            participants = study_mapped_df_dict["participants"]
+            participants = participants.rename(
```

Comment:
Should this also have a None check like the other ones?
```suggestion
            if participants is not None:
                participants = participants.rename(
```

Reply:
Given this FHIR ETL runs against a "released" study(ies), there are a few endpoints that always exist (confirmed with DataOps as well):

- `/studies`
- `/participants`
- `/biospecimens`
- `/biospecimen-genomic-files`
- `/genomic-files`

Except for these endpoints, I've set the rest as conditional so that the ETL won't crash.

Reply:
So initially, I thought this was ok, but as I try to run this locally I'm finding it to be a hinderance. I think it would be nice if we didn't assume data exists at all of these endpoints for multiple reasons:

- We might want to run the ETL for a study that doesn't have all of its data yet but we still want to see what it looks like in the FHIR server
- A developer may want to run the ETL locally for testing/development purposes and it would be nice if they could run it on a simple study with just participants

Reply:
I made them all conditional.